Policy support for replicators, pods, nodes

cben opened the pull request:
Completes usable subset of #8654.
https://bugzilla.redhat.com/show_bug.cgi?id=1341253 (master)
https://bugzilla.redhat.com/show_bug.cgi?id=1346057 (darga)
[This PR replaces #9538, #8815.]

- Doesn't show compliance status/history on these entities.  (will come in separate PR)
- Didn't touch alerts, simulation.

Depends on #9601 for the events to arrive, ~~but split for independent review~~ now rebased after #9601 got merged.
Would benefit from #9347 (to say "Pod" instead of "Container Group" in a few places) but not functionally depends on it.

Heavy cargo-cult of Mooli's #7132 and other PRs (thanks!).

- [x] Policies UI:
  - [x] Can create & edit Control & Compliance policies.
  - [x] Can add replicator/pod/node events.
  - [x] Can add conditions, and actions.
- [x] Can add the new policies to a Policy Profile and have them run.
  - [x] Assigning the profile to the EMS works.
  - [x] Assigning to specific replicator/pod/node works.
- [x] Event propogation works end-to-end.
  - [x] Incoming POD_FAILEDSYNC events trigger a policy triggering a "Generate log" action.
  - [x] Incoming NODE_REBOOTED event triggers a policy triggering a "Generate log" action.
  - [x] Incoming REPLICATOR_SUCCESSFULCREATE event triggers a policy triggering "Generate log" action.
- [x] Compliance:
  - [x] "Check Compliance" buttons in UI runs compliance policies.
  - [x] "Mark as non-compliant" action work from a compliance policy.
  - [ ] "Mark as non-compliant" action from *control* policy: executed according to policy.log, however `Compliance` table doesn't reflect that.  Not sure it's supported in Control Policies?  Shouldn't block this PR.

### policy.log showing it all works (with #9601)

https://gist.github.com/cben/870d6fab478963ab9e78cbfa776d3194

### screenshots 
(See also #9601 for Events list and policy event assignment screenshots.  The screenshots here were made with #9601.)

![screenshot-localhost 3000 2016-07-17 14-17-41 policies](https://cloud.githubusercontent.com/assets/273688/16900749/131fe482-4c37-11e6-88d9-52f44aba4db7.png)
![screenshot-localhost 3000 2016-07-17 14-19-18 policy profile](https://cloud.githubusercontent.com/assets/273688/16900759/4ce2d86e-4c37-11e6-9bd5-f3c86bea7194.png)
![screenshot-localhost 3000 2016-07-17 13-20-02 replicant compliance condition](https://cloud.githubusercontent.com/assets/273688/16900761/525e852c-4c37-11e6-95c4-61d6fd6ea0f0.png)
![multi pod assign](https://cloud.githubusercontent.com/assets/273688/16900766/73fc9c14-4c37-11e6-994f-90d720f37e33.png)

miq-bot add-labels control, ui, providers/containers, enhancement, darga/yes, ~~wip~~

martinpovolny commented on app/helpers/application_helper/toolbar/container_group_center.rb:
```diff
@@ -58,6 +58,17 @@ class ApplicationHelper::Toolbar::ContainerGroupCenter < ApplicationHelper::Tool
+        button(
+          :container_group_protect,
+          'pficon pficon-edit fa-lg',
+          N_('Manage Policies for this #{ui_lookup(:table=>"container_group")}'),
```
this will not work any more.

You need a proc if you want to do interpolations. And the preferred way it to avoid the `ui_lookup`. Just put "Container Group" or whatever it is called in.

cben replied:
Thanks!  Done X 6.

martinpovolny commented on app/helpers/application_helper/toolbar/container_nodes_center.rb:
```diff
@@ -47,6 +47,23 @@ class ApplicationHelper::Toolbar::ContainerNodesCenter < ApplicationHelper::Tool
+        button(
+          :container_node_protect,
+          'pficon pficon-edit fa-lg',
+          N_('Manage Policies for this #{ui_lookup(:table=>"container_nodes")}'),
```
same here

martinpovolny commented on app/helpers/application_helper/toolbar/container_replicator_center.rb:
```diff
@@ -58,6 +58,17 @@ class ApplicationHelper::Toolbar::ContainerReplicatorCenter < ApplicationHelper:
+        button(
+          :container_replicator_protect,
+          'pficon pficon-edit fa-lg',
+          N_('Manage Policies for this #{ui_lookup(:table=>"container_replicator")}'),
```
and here

martinpovolny commented on app/helpers/application_helper/toolbar/container_replicators_center.rb:
```diff
@@ -47,6 +47,23 @@ class ApplicationHelper::Toolbar::ContainerReplicatorsCenter < ApplicationHelper
+        button(
+          :container_replicator_protect,
+          'pficon pficon-edit fa-lg',
+          N_('Manage Policies for this #{ui_lookup(:table=>"container_replicators")}'),
```
here

simon3z commented:
@miq-bot assign cben

cben commented:
## Ready for review

@miq-bot remove-label wip

I'll address rubocop tomorrow.

moolitayer commented on app/controllers/application_controller/tree_support.rb:
```diff
@@ -22,11 +22,15 @@ def squash_toggle
+            else
+              VmOrTemplate
+            end
+    identify_record(params[:id], klass)
```
I prefer the existing syntax better (mostly because it is a plain string comparison as appose to === done by case)

cben replied:
Done.

moolitayer commented on app/controllers/miq_policy_controller/policies.rb:
```diff
@@ -226,7 +226,8 @@ def excluded_event?(event)
   def policy_get_all_folders(parent = nil)
     if !parent.nil?
-      @folders = MiqPolicy::UI_FOLDERS.collect(&:name)
+      # TODO: this is dead code, the one caller passes no arguments
```
I agree. Could you remove the todo and submit a separate pr removing the dead code?
It isn't urgent but would be nice (please do not depend on that pr if you do)

cben replied:
https://github.com/ManageIQ/manageiq/compare/master...cben:dead-code-in-policy_get_all_folders
will open the PR after this gets merged.

moolitayer commented on app/models/ems_event.rb:
```diff
@@ -338,6 +338,29 @@ def ems_refresh_target
+
+  def src_container_node
+    # TODO: resolve functional duplication with process_container_entities_in_event!
+    #       (https://github.com/ManageIQ/manageiq/issues/9600)
```
are the above two different items? is the query below fixing #9600?

cben replied:
The query below is working around #9600; once `process_container_entities_in_event!` is fixed, this should use `container_node_id`.

Doing a similar query in `process_container_entities_in_event!` would "fix" #9600.
But it's not as "reliable" as a real uid, if we can make k8s/kubeclient send it.  Might need to fall back on a name query when uid is missing.  Didn't want to block this PR on figuring out the right fix...

moolitayer commented:
@cben I see this is dependent upon #9601 I thought we wanted to avoid dependencies so that we can have both reviewed. If it is simple please separate if not let's discuss it.

cben commented on app/controllers/application_controller/tree_support.rb:
```diff
@@ -22,11 +22,15 @@ def squash_toggle
-      identify_record(params[:id], VmOrTemplate)
-    end
+    # TODO: This logic should probably be reversed - fixed list for VmOrTemplate.
+    # (Better yet, override the method only in VmOrTemplate related controllers.)
```
The reason I'm not doing this is I'm not sure what's the exact list that should result in `VmOrTemplate`.
The original `controller_name == 'host' ? Host : VmOrTemplate` code dates back to the "Initial Commit", 
and `find_record` seems widely used.

moolitayer commented:
:+1: Once the dependency on #9601 is broken (most of my comments where given on #8815)

@gmcculloug @gtanzillo Could one of you please assign reviewers for this pr?
we would like to have this by the code freeze

cben commented:
Done, split from #9601. (There will be mild merge conflicts, when one gets merged I'll rebase the other.)

simon3z commented:
@moolitayer can you review this? Also please add some other reviewers that could help here. Thanks.

moolitayer commented:
> @moolitayer can you review this? Also please add some other reviewers that could help here. Thanks.

LGTM :+1: see [comment](https://github.com/ManageIQ/manageiq/pull/9813#issuecomment-233281369) 

@cben 
- thanks for breaking the dependency 
- please make sure this starts being reviewed today. Either @gmcculloug or @gtanzillo should assign reviewers (I'm not sure which team is currently covering control  but i'll find out soon)

gtanzillo commented:
👍 LGTM. Since the changes are mostly UI related I'd like to wait for +1 from someone on that team ( @martinpovolny @dclarizio) before merging.

cben commented:
@martinpovolny @dclarizio could someone review this from UI perspective?

dclarizio commented:
@martinpovolny @h-kataria please try to review this tomorrow, if possible.  Thx, Dan

dclarizio commented:
@cben need to rebase/rebush.

h-kataria commented:
Looks good, most of the changes are related to adding new toolbar buttons, product features and deals with showing new tree nodes for Containers related CIs.

dclarizio commented:
@cben I did notice this discrepancy in the Policies and Conditions `Pod` areas.  In the tree I see Pods, but on the buttons and in the right side, it shows Container Groups:
![pod conditions](https://cloud.githubusercontent.com/assets/3433754/17063911/9222692a-4fee-11e6-96b6-36a51def92d7.png)

dclarizio commented:
@cben Also, comparing Policies accordion to the Conditions accordion above, one says `Container Replicator Conditions`, the other says `Replicator Policies` (no Container prefix):
![replicators](https://cloud.githubusercontent.com/assets/3433754/17064013/267569ec-4fef-11e6-998d-37752611213e.png)

dclarizio commented:
@cben seeing some images missing, for example:
`ActionController::RoutingError (No route matches [GET] "/images/100/event-containerreplicator_compliance_check.png"):`
If you can get a list, @epwinchell can get some added either in this PR or a separate one.

cben commented:
Rebased (yay, no more `expected_features_count`)

@dclarizio
> I did notice this discrepancy in the Policies and Conditions Pod areas. In the tree I see Pods, but on the buttons and in the right side, it shows Container Groups

This, at least right side, is result of failed `ui_lookup` due to wrong case.
#9347 fixes these.  I closed it as @martinpovolny objected to fixing ui_lookup rather than removing it, but it can easily be revivied.

- [ ] I'll check where the buttons happen.

> Also, comparing Policies accordion to the Conditions accordion above, one says Container Replicator Conditions, the other says Replicator Policies (no Container prefix)

Thanks, fixed to "Replicator", as used in most places.

![screenshot-localhost 3001 2016-07-24 12-04-33 conditions](https://cloud.githubusercontent.com/assets/273688/17082911/e6b298a8-5196-11e6-9edd-f0998c13b1ab.png)

> seeing some images missing, for example:
ActionController::RoutingError (No route matches [GET] "/images/100/event-containerreplicator_compliance_check.png"):

Weird.  Many event images would be missing if you tested without #9601, before it got merged and I rebased this PR.  But specifically
`app/assets/images/100/event-containerreplicator_compliance_check.png`
is added by this PR (b6fec06689cb9b01398855cf3553f6ea19ac02a6).
I see it shown here:
![screenshot-localhost 3001 2016-07-24 12-18-58 events](https://cloud.githubusercontent.com/assets/273688/17082975/002044f0-5199-11e6-8529-a78525f7fd18.png)
but that uses `/assets/100/event-containerreplicator_compliance_check-8c6dd6af5a7d4d9795b9ca186a603d34a6fbcaa25f774550f687458788ec321d.png` URL.
So do several other places I now tried that show this image.

Any idea where `/images/100/event-containerreplicator_compliance_check.png` URL (which indeed 404s) could be requested?

cben commented:
Oops!  Top-level Policies->Compliance/Control Policies lists are broken.
```
Error caught: [NameError] uninitialized constant MiqPolicyController::POLICY_FOLDERS
/home/bpaskinc/miq/events/app/controllers/miq_policy_controller.rb:943:in `folder_get_info'
/home/bpaskinc/miq/events/app/controllers/miq_policy_controller.rb:447:in `get_node_info'
/home/bpaskinc/miq/events/app/controllers/miq_policy_controller.rb:292:in `tree_select'
```
Remanant from when I was trying to do a PR without compliance.  Fixed, verified it doesnt't appear elsewhere.
This still leaves missing right-side labels:
![screenshot-localhost 3001 2016-07-24 13-23-33 policies empty labels](https://cloud.githubusercontent.com/assets/273688/17083244/e6cf18c4-51a1-11e6-9c43-e31d4af1f2f5.png)

=> I need to add some strings to _policy_folders.html.haml due to #9543 which was darga/no, so preparing a separate darga/no PR for that...

cben commented:
Done.  Verified that on darga with this PR but without #10013, these lists will function, with slightly suboptimal (and non-translatable) text:

![screenshot-localhost 3001 2016-07-24 13-23-33 policies darga labels](https://cloud.githubusercontent.com/assets/273688/17083392/4b6f2ade-51a8-11e6-8292-b9860a696235.png)

If it's desirable to fix, just reconsider backporting #9543 + #10013.  I'm OK either way.

@dclarizio Believe I've addressed everything.

moolitayer commented:
@dclarizio @mzazrivec Is it safe to backport #9543?

@cben I thought I read in #10013 That its only for master?
Is the patch relevant for Darga?

cben commented:
#10013's darga status should match #9543's darga status.
Both backported or neither.  Can be decided orthogonally to this PR.

moolitayer commented:
Thanks for the clarification @cben

simon3z commented:
Ping? Ready for merge :+1: 

cc @chessbyte @blomquisg

simon3z commented:
@cben @moolitayer where are the specs?

chessbyte commented:
Darga backport details:
```text
commit 27f82103b80b0e2c0dac63efbff8373f340d41c0
Author: Oleg Barenboim <chessbyte@gmail.com>
Date:   Mon Jul 25 14:09:40 2016 -0400

    Merge pull request #9813 from cben/container-policies
    
    Policy support for replicators, pods, nodes
    (cherry picked from commit 773ebeee5bd22c6c4664a48999827f658463a6bd)
```
